setup.py

```diff
@@ -744,6 +744,7 @@ def build_mx_extension(build_ext, options):
     else:
         mxnet_mpi_lib.define_macros += [('MXNET_USE_MKLDNN', '0')]
     mxnet_mpi_lib.define_macros += [('MSHADOW_USE_MKL', '0')]
+    mxnet_mpi_lib.define_macros += [('MSHADOW_USE_F16C', '0')]
```

Comment:
We already set it to '0', which means disable. How did it cause installation failure?

Reply:
It was not set to '0'. This flag is not used in horovod anyways. Even in MXNet, it should not be set to '1'. See https://github.com/dmlc/mshadow/blob/1d79ecfdb4c9234537e1bf5148f44a1af54501ec/make/mshadow.mk#L61

Reply:
It caused installation failure due to https://github.com/dmlc/mshadow/blob/1d79ecfdb4c9234537e1bf5148f44a1af54501ec/mshadow/half.h#L79

Reply:
This is set automatically by detecting if the platform supports f16c instructions, right? https://github.com/dmlc/mshadow/blob/1d79ecfdb4c9234537e1bf5148f44a1af54501ec/make/mshadow.mk#L36-L56 

I did find f16c on DL AMI 18.0 (AL): `cat /proc/cpuinfo | grep flags | grep f16c`
```
fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single pti fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx xsaveopt
```

Reply:
Also if hvd-mxnet does automatic gradient compression in the future, f16c will help accelerate fp32<->fp16 conversion, I assume?

Reply:
BTW - I find horovod 0.16.2 installation is fine on DLAMI. But 0.16.3 fails

Reply:
@e

> This is set automatically by detecting if the platform supports f16c instructions, right? https://github.com/dmlc/mshadow/blob/1d79ecfdb4c9234537e1bf5148f44a1af54501ec/make/mshadow.mk#L36-L56
> 
> I did find f16c on DL AMI 18.0 (AL): `cat /proc/cpuinfo | grep flags | grep f16c`
> 
> ```
> fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single pti fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx xsaveopt
> ```

The line you pasted is set when building MXNet with USE_F16C flag. In this case, the flag is to actually link a header file when building Horovod with MXNet as the dependency. It is set here: https://github.com/dmlc/mshadow/blob/6e94643bdf1d51a505b147f28c358fb71070b8fd/mshadow/base.h#L139

In either case, that flag is set automatically when building libmxnet.so. It seems we do not need to turn on the compiler flag when building Horovod.